File: customer/src/main/java/br/com/accenture/wallet/customer/service/CustomerService.java
package br.com.accenture.wallet.customer.service;

import br.com.accenture.wallet.customer.domain.Customer;
import br.com.accenture.wallet.customer.domain.CustomerModel;
import br.com.accenture.wallet.customer.repository.CustomerRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
public class CustomerService {

    private CustomerRepository repository;

    @Autowired
    public CustomerService(CustomerRepository repo) {
        this.repository = repo;
    }

    public CustomerModel create(final CustomerModel model) {
        Customer customer = new Customer();
        model.setId(customer.getId());
        repository.save(model.toEntity());
        return model;
    }

    public CustomerModel update(final CustomerModel model) {
        final Optional<Customer> customer = repository.findById(model.getId());
        if (customer.isEmpty()) return null;
        repository.save(model.toEntity());
        return model;
    }

    public void remove(String id) {
        repository.deleteById(id);
    }

    public CustomerModel findById(String value) {
        Optional<Customer> customer = repository.findById(value);
        if (customer.isEmpty()) return null;
        return new CustomerModel("", "").fromEntity(customer.get());
    }

    public List<CustomerModel> findAll() {
        return repository.findAll()
            .stream()
            .map(customer -> new CustomerModel("", "").fromEntity(customer))
            .collect(Collectors.toList());
    }

    public CustomerModel findByEmail(String value) {
        final Optional<Customer> customer = repository.findByEmail(value);
        if (customer.isEmpty()) return null;
        return new CustomerModel("", "").fromEntity(customer.get());
    }

}

File: customer/src/main/kotlin/br/com/accenture/wallet/customer/service/CustomerService.kt
package br.com.accenture.wallet.customer.service

import br.com.accenture.wallet.customer.domain.CustomerBean
import br.com.accenture.wallet.customer.repository.CustomerRepository
import org.springframework.beans.factory.config.ConfigurableBeanFactory
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Service
import java.util.*
import java.util.stream.Collectors
import java.util.Objects.isNull

@Service
@Scope(ConfigurableBeanFactory.SCOPE_SINGLETON)
class CustomerService(val repository: CustomerRepository) {

    fun create(bean: CustomerBean): CustomerBean {
        val customer = bean.toEntity()
        customer.id = UUID.randomUUID().toString()
        return bean.fromEntity(repository.save(customer))
    }

    fun remove(id: String) {
        repository.deleteById(id)
    }

    fun update(model: CustomerBean): CustomerBean? {
        val customer = repository.findById(model.id)
        if (customer.isEmpty) return null
        repository.save(model.toEntity())
        return model
    }

    fun findById(value: String): Optional<CustomerBean> {
        val customer = repository.findById(value)
        return if (customer.isEmpty) Optional.empty() else Optional.of(CustomerBean().fromEntity(customer.get()))
    }

    fun findAll(): List<CustomerBean> {
        return repository.findAll()
            .stream()
            .map { customer -> CustomerBean().fromEntity(customer) }
            .collect(Collectors.toList())
    }

    fun findByEmail(value: String?): CustomerBean? {
        if (isNull(value)) return null
        val customer = repository.findByEmail(value!!)
        return if (customer.isEmpty) null else CustomerBean().fromEntity(customer.get())
    }

}
